File: ReverseWithTag/ReverseWithTag.scala
object ReverseWithTag {
  private def tokenize(text: String): List[String] = {
    if (text isEmpty) List()
    else {
      val tagStart = text.indexOf('<')
      tagStart match {
        case -1 => List(text)
        case 0 => {
          val tagEnd = text.indexOf('>')
          text.take(tagEnd + 1) :: tokenize(text.drop(tagEnd + 1))
        }
        case _ => text.take(tagStart) :: tokenize(text.drop(tagStart))
      }
    }
  }

  def reverseTokens(tokens: List[String]): List[String] = {
    def isTag(token: String) = { token(0) == '<' }
    def isCloseTag(token: String) = { isTag(token) && token(1) == '/' }

    def tagName(tag: String) = {
      val inside =
        if (isCloseTag(tag)) tag.drop(2).dropRight(1)
        else tag.drop(1).dropRight(1)
      inside.split(' ').head
    }

    def findCloseTag(index: Int, name: String): Int = {
      if (tokens isEmpty) 0
      else if (isCloseTag(tokens(index)) && tagName(tokens(index)) == name) index
      else findCloseTag(index + 1, name)
    }

    if (tokens isEmpty) List()
    else if (isTag(tokens.head)) {
      val closeIndex = findCloseTag(1, tagName(tokens.head))
      val insideElement = tokens.slice(1, closeIndex)
      val afterElement = tokens.drop(closeIndex + 1)

      reverseTokens(afterElement) ::: tokens(0) ::
      reverseTokens(insideElement) ::: tokens(closeIndex) :: Nil
    } else {
      reverseTokens(tokens.tail) :::
      tokens.head.reverse :: Nil
    }
  }

  def reverseText(text: String) = {
    val tokens = tokenize(text)
    reverseTokens(tokens).mkString
  }

  def main(args: Array[String]) = {
    println(reverseText("Hello <b attr='some'>Bold <i>Italic</i> words</b> World"))
  }
}

File: ReverseWithTag/ReverseWithTag.js
/**
 * Tokenize text into Array(String)
 */
function tokenize(text) {
	if (text == '') return [];

	var startIndex = text.indexOf('<');
	var endIndex = text.indexOf('>');

	if (startIndex < 0) return [text];
	else if (startIndex == 0 ) {
		return [text.slice(0, endIndex + 1)].concat(
			tokenize(text.slice(endIndex + 1, text.length))
		);
	}

	return [text.slice(0, startIndex)].concat(
		tokenize(text.slice(startIndex, text.length))
	);
}

// Utility functions
function isTag(token) { return token[0] == '<'; }
function isCloseTag(token) { return isTag(token) && token[1] == '/'; }
function tagName(token) {
	var inside = isCloseTag(token) ? token.slice(2, -1) : token.slice(1, -1);
	return inside.split(' ')[0];
}

/**
 * Reverse tokens with tag preserved
 */
function reverseTokens(tokens) {
	function findCloseTag(index, name) {
		// Didn'to check error case
		if (tokens[index].indexOf("</" + name) == 0) return index;
		return findCloseTag(index + 1, name);
	}

	if (tokens.length == 0) return [];
	if (isTag(tokens[0])) {
		var closeIndex = findCloseTag(1, tagName(tokens[0]));
		var insideElement = tokens.slice(1, closeIndex);
		var afterElement = tokens.slice(closeIndex + 1, tokens.length);

		return reverseTokens(afterElement).concat([tokens[0]]).
			concat(reverseTokens(insideElement)).concat([tokens[closeIndex]]);
	}

	return reverseTokens(tokens.slice(1)).concat([tokens[0].split('').reverse().join('')]);
}

/**
 * Reverse text with tag preserved
 */
function reverse(text) {
	return reverseTokens(tokenize(text)).join('');
}


// ----------------------------------------------------------------
// Test
// ----------------------------------------------------------------
var origin = "Hello <b attr='some'>Bold <i>Italic</i> words</b> World";
//
//var tokens = tokenize(origin);
//tokens.forEach(function(token) {
//	console.log('Token [' + token + '] is ' +
//		(isTag(token) ? (isCloseTag(token) ? 'Close' : 'Open') + ' Tag' : 'Word')
//	);
//	if (isTag(token)) console.log('Name: ', tagName(token));
//});
//
var reverseText = reverse(origin);
console.log(reverseText);

